// v-gsap-nuxt/dist/module.mjs
import { defineNuxtModule, createResolver, addPlugin, addImports, addComponent } from '@nuxt/kit';
import { defu } from 'defu';

const module = defineNuxtModule({
  meta: {
    name: "v-gsap-nuxt",
    configKey: "vgsap",
    compatibility: {
      nuxt: ">=3.0.0"
    }
  },
  // Default configuration options of the Nuxt module
  defaults: {},
  setup(_options, _nuxt) {
    const resolver = createResolver(import.meta.url);
    _nuxt.options.runtimeConfig.public.vgsap = defu(
      _nuxt.options.runtimeConfig.public.vgsap,
      {
        ..._options
      }
    );
    _nuxt.options.css.push(resolver.resolve("./runtime/styles/vgsap.css"));
    addPlugin(resolver.resolve("./runtime/nuxt"));
    if (_nuxt.options.runtimeConfig.public.vgsap?.composable != false) {
      addImports({
        name: "useGSAP",
        as: "useGSAP",
        from: resolver.resolve("runtime/plugin")
        // load composable from plugin
      });
    }
    addComponent({
      name: "GSAPTransition",
      filePath: resolver.resolve("./runtime/components/GSAPTransition.vue")
    });
  }
});

export { module as default };
